File: ui-tests/PageObjects/Panes/Scratchpad/ScratchpadPane.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects;
using UiTests.Utils;

namespace UiTests.PageObjects.Panes.Scratchpad;

/// <summary>
/// Page object wrapping the scratchpad pane and its stored values.
/// </summary>
public class ScratchpadPane : TabObject
{
    private List<ScratchpadEntry> _entries = new();

    public ScratchpadPane(IPage page, ILocator root, string tabButtonText)
        : base(page, root, tabButtonText)
    {
    }

    /// <summary>
    /// Locator targeting the container that renders scratchpad values.
    /// </summary>
    public ILocator EntriesContainer()
        => Root.Locator(".value-components-container");

    /// <summary>
    /// Retrieves the currently visible scratchpad entries.
    /// </summary>
    public async Task<IReadOnlyList<ScratchpadEntry>> EntriesAsync(bool forceReload = false)
    {
        if (forceReload || _entries.Count == 0)
        {
            var entryRoots = await EntriesContainer()
                .Locator(".scratchpad-value-view")
                .AllAsync();
            _entries = entryRoots
                .Select(locator => new ScratchpadEntry(locator))
                .ToList();
        }

        return _entries;
    }

    /// <summary>
    /// Returns the number of entries currently displayed.
    /// </summary>
    public Task<int> EntryCountAsync()
        => EntriesContainer().Locator(".scratchpad-value-view").CountAsync();

    /// <summary>
    /// Waits until the scratchpad renders at least <paramref name="count"/> entries.
    /// </summary>
    public Task WaitForEntryCountAsync(int count)
        => RetryHelpers.RetryAsync(async () => await EntryCountAsync() >= count);

    /// <summary>
    /// Waits for a new entry to appear based on a previous baseline.
    /// </summary>
    public Task WaitForNewEntryAsync(int previousCount)
        => WaitForEntryCountAsync(previousCount + 1);

    /// <summary>
    /// Locates the first entry whose expression label matches <paramref name="expression"/>.
    /// Returns <c>null</c> when no such entry exists.
    /// </summary>
    public async Task<ScratchpadEntry?> FindEntryAsync(string expression, bool forceReload = false)
    {
        var entries = await EntriesAsync(forceReload);
        foreach (var entry in entries)
        {
            var expr = await entry.ExpressionAsync();
            if (expr == expression)
            {
                return entry;
            }
        }

        return null;
    }

    /// <summary>
    /// Computes the current entries keyed by their expression label.
    /// </summary>
    public async Task<Dictionary<string, ScratchpadEntry>> EntryMapAsync(bool forceReload = false)
    {
        var map = new Dictionary<string, ScratchpadEntry>(StringComparer.OrdinalIgnoreCase);
        var entries = await EntriesAsync(forceReload);
        foreach (var entry in entries)
        {
            var expr = await entry.ExpressionAsync();
            map[expr] = entry;
        }

        return map;
    }

    /// <summary>
    /// Clears the cached entry list, forcing a fresh query on the next access.
    /// </summary>
    public void InvalidateCache()
    {
        _entries.Clear();
    }

    /// <summary>
    /// Waits until an entry matching <paramref name="expression"/> appears and returns it.
    /// </summary>
    public async Task<ScratchpadEntry> WaitForEntryAsync(string expression)
    {
        ScratchpadEntry? found = null;
        await RetryHelpers.RetryAsync(async () =>
        {
            InvalidateCache();
            found = await FindEntryAsync(expression, forceReload: true);
            return found is not null;
        });

        return found!;
    }
}

File: tsc-ui-tests/page-objects/panes/scratchpad/scratchpad-pane.ts
import type { Locator, Page } from "@playwright/test";
import { ScratchpadEntry } from "./scratchpad-entry";
import { retry } from "../../../lib/retry-helpers";

/**
 * Page object wrapping the scratchpad pane and its stored values.
 *
 * Port of ui-tests/PageObjects/Panes/Scratchpad/ScratchpadPane.cs
 */
export class ScratchpadPane {
  readonly page: Page;
  readonly root: Locator;
  readonly tabButtonText: string;
  private entries_: ScratchpadEntry[] = [];

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
  }

  tabButton(): Locator {
    return this.page.locator(".lm_title", { hasText: this.tabButtonText }).first();
  }

  async isVisible(): Promise<boolean> {
    const style = await this.root.locator("..").getAttribute("style");
    return !(style?.includes("none"));
  }

  entriesContainer(): Locator {
    return this.root.locator(".value-components-container");
  }

  async entries(forceReload = false): Promise<ScratchpadEntry[]> {
    if (forceReload || this.entries_.length === 0) {
      const roots = await this.entriesContainer()
        .locator(".scratchpad-value-view")
        .all();
      this.entries_ = roots.map((l) => new ScratchpadEntry(l));
    }
    return this.entries_;
  }

  async entryCount(): Promise<number> {
    return this.entriesContainer().locator(".scratchpad-value-view").count();
  }

  async waitForEntryCount(count: number): Promise<void> {
    await retry(async () => (await this.entryCount()) >= count);
  }

  async waitForNewEntry(previousCount: number): Promise<void> {
    await this.waitForEntryCount(previousCount + 1);
  }

  async findEntry(
    expression: string,
    forceReload = false,
  ): Promise<ScratchpadEntry | null> {
    const entries = await this.entries(forceReload);
    for (const entry of entries) {
      const expr = await entry.expression();
      if (expr === expression) {
        return entry;
      }
    }
    return null;
  }

  async entryMap(forceReload = false): Promise<Map<string, ScratchpadEntry>> {
    const map = new Map<string, ScratchpadEntry>();
    const entries = await this.entries(forceReload);
    for (const entry of entries) {
      const expr = await entry.expression();
      map.set(expr, entry);
    }
    return map;
  }

  invalidateCache(): void {
    this.entries_ = [];
  }

  async waitForEntry(expression: string): Promise<ScratchpadEntry> {
    let found: ScratchpadEntry | null = null;
    await retry(async () => {
      this.invalidateCache();
      found = await this.findEntry(expression, true);
      return found !== null;
    });
    return found!;
  }
}
